js/app.js

```diff
@@ -110,20 +111,41 @@ define([
 					},					
 				};
 
-                const asyncBefore = function() {
-                	// Refresh auth token
-                	return (self.config.userAuthenticationEnabled && authApi.token() != null && authApi.tokenExpirationDate() > new Date()) ? authApi.refreshToken() : new Promise(resolve => resolve());
-                };
-
-                // Since the router's "before" hook doesn't work with promises,
+				let onLoginSubscription;
+				self.activeRouteHandler = null;
+				// Since the router's "before" hook doesn't work with promises,
 				// there is a need to wrap the routes manually
-                const routesWithRefreshedToken = Object.keys(routes).reduce((accumulator, key) => {
-                	accumulator[key] = function() {
-                        self.currentView('loading');
-                		asyncBefore().then(() => routes[key].apply(null, arguments));
-                    };
-                	return accumulator;
+				const routesWithRefreshedToken = Object.keys(routes).reduce((accumulator, key) => {
+					accumulator[key] = function() {
+						self.currentView('loading');
+						if (onLoginSubscription) {
+							onLoginSubscription.dispose();
+						}
+						const handler = routes[key].handler.bind(null, ...arguments);
+						routes[key].checkPermission()
+							.then(() => handler())
+							.catch(() => {
+								// protected route didn't pass the token check -> show white page
+								self.currentView('white-page');
+								// wait until user authenticates
+								onLoginSubscription = authApi.isAuthenticated.subscribe((isAuthenticated) => {
+									if (isAuthenticated) {
+										handler();
+										onLoginSubscription.dispose();
+									}
+								});
+							});
+
+						self.activeRouteHandler = arguments.callee.bind(this, ...arguments);
+					};
+					return accumulator;
 				}, {});
+				// anyway, we should track the moment when the user exits and check permissions once again
+				authApi.isAuthenticated.subscribe((isAuthenticated) => {
+					if (!isAuthenticated) {
+						self.activeRouteHandler();
```

Comment:
1) what's this doing?
2) cannot we face situation, when the `activeRouteHandler` will be `undefined`?

Reply:
1. Hanldes the situation when you click 'log out' on the protected page and activates the white page 
2. No, because the router will handle one of the routes anyway. Then activeRouteHandler will be set to a function